core/shared/src/main/scala/zio/internal/macros/Graph.scala

```diff
@@ -25,8 +25,16 @@ final case class Graph[Key, A](nodes: List[Node[Key, A]], keyEquals: (Key, Key)
   def map[B](f: A => B): Graph[Key, B] =
     Graph(nodes.map(_.map(f)), keyEquals)
 
-  private def getNodeWithOutput[E](output: Key, error: E): Either[::[E], Node[Key, A]] =
-    nodes.find(_.outputs.exists(keyEquals(_, output))).toRight(::(error, Nil))
+  private val nodeWithOutputCache = scala.collection.mutable.Map.empty[Key, Option[Node[Key, A]]]
+
+  private def getNodeWithOutput[E](output: Key, error: => E): Either[::[E], Node[Key, A]] =
+    (if (nodeWithOutputCache.contains(output)) {
+       nodeWithOutputCache.apply(output)
+     } else {
+       val n = nodes.find(_.outputs.exists(keyEquals(_, output)))
+       nodeWithOutputCache.put(output, n)
+       n
+     }).toRight(::(error, Nil))
```

Comment:
Very minor, but how about this? I think it reads a bit better

```suggestion
  private def getNodeWithOutput[E](output: Key, error: => E): Either[::[E], Node[Key, A]] =
    nodeWithOutputCache
      .updateWith(output) {
        case s @ Some(_) => s
        case None        => Some(nodes.find(_.outputs.exists(keyEquals(_, output))))
      }
      .get
      .toRight(::(error, Nil))
```

Reply:
I wrote it like that for efficiency, but if I use a Java `HashMap` I can get both and nicer readability.